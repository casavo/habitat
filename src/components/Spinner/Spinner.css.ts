/* eslint-disable sort-keys-fix/sort-keys-fix */
import { keyframes } from "@vanilla-extract/css";
import { recipe } from "@vanilla-extract/recipes";
import { vars } from "../../utils/theme.css";

const sizes: Record<"large" | "medium" | "small", number> = {
  large: 28,
  medium: 24,
  small: 20,
};

const rotate = keyframes({
  "0%": { transform: "rotate(0deg)" },
  "100%": { transform: "rotate(360deg)" },
});

export const Container = recipe({
  base: {
    position: "relative",
    display: "inline-block",
  },
  variants: {
    size: {
      small: {
        height: 20,
        width: 20,
        position: "relative",
      },
      medium: {
        height: 24,
        width: 24,
        position: "relative",
      },
      large: {
        height: 28,
        width: 28,
        position: "relative",
      },
    },
  },
});

export const Circle = recipe({
  base: {
    border: `1px solid`,
    borderRadius: "100%",
    boxSizing: "border-box",
    display: "block",
    position: "absolute",
  },

  variants: {
    theme: {
      dark: {},
      light: {},
    },
    variant: {
      primary: {},
      secondary: {},
      tertiary: {},
    },
    size: {
      small: {
        height: sizes["small"] - 1,
        width: sizes["small"] - 1,
        lineHeight: sizes["small"],
        left: `calc(50% - ${(sizes["small"] - 1) / 2}px)`,
        top: `calc(50% - ${(sizes["small"] - 1) / 2}px)`,
      },
      medium: {
        height: sizes["medium"] - 1,
        width: sizes["medium"] - 1,
        lineHeight: sizes["medium"] - 1,
        left: `calc(50% - ${(sizes["medium"] - 1) / 2}px)`,
        top: `calc(50% - ${(sizes["medium"] - 1) / 2}px)`,
      },
      large: {
        height: sizes["large"] - 1,
        width: sizes["large"] - 1,
        lineHeight: sizes["large"],
        left: `calc(50% - ${(sizes["large"] - 1) / 2}px)`,
        top: `calc(50% - ${(sizes["large"] - 1) / 2}px)`,
      },
    },
  },
  compoundVariants: [
    {
      variants: {
        theme: "dark",
        variant: "primary",
      },
      style: {
        borderColor: vars.colors.root.neutral[0],
      },
    },
    {
      variants: {
        theme: "dark",
        variant: "secondary",
      },
      style: {
        borderColor: vars.colors.root.neutral[600],
      },
    },
    {
      variants: {
        theme: "light",
        variant: "primary",
      },
      style: {
        borderColor: vars.colors.root.neutral[600],
      },
    },
    {
      variants: {
        theme: "light",
        variant: "secondary",
      },
      style: {
        borderColor: vars.colors.root.neutral[0],
      },
    },
  ],
});

export const Spin = recipe({
  base: {
    animation: `${rotate} 0.6s ease-out infinite`,
    border: `2px solid`,
    borderColor: "transparent",
    borderRadius: "100%",
    borderTop: `2px solid`,
    boxSizing: "border-box",
    display: "block",
    position: "absolute",
  },

  variants: {
    variant: {
      primary: {},
      secondary: {},
      tertiary: {},
    },
    theme: {
      dark: {},
      light: {},
    },
    size: {
      small: {
        height: sizes["small"] - 1,
        width: sizes["small"] - 1,
        lineHeight: sizes["small"],
        left: `calc(50% - ${(sizes["small"] - 1) / 2}px)`,
        top: `calc(50% - ${(sizes["small"] - 1) / 2}px)`,
      },
      medium: {
        height: sizes["medium"] - 1,
        width: sizes["medium"] - 1,
        lineHeight: sizes["medium"],
        left: `calc(50% - ${(sizes["medium"] - 1) / 2}px)`,
        top: `calc(50% - ${(sizes["medium"] - 1) / 2}px)`,
      },
      large: {
        height: sizes["large"] - 1,
        width: sizes["large"] - 1,
        lineHeight: sizes["large"],
        left: `calc(50% - ${(sizes["large"] - 1) / 2}px)`,
        top: `calc(50% - ${(sizes["large"] - 1) / 2}px)`,
      },
    },
  },
  compoundVariants: [
    {
      variants: {
        theme: "dark",
        variant: "primary",
      },
      style: {
        borderTopColor: vars.colors.root.neutral[0],
      },
    },
    {
      variants: {
        theme: "dark",
        variant: "secondary",
      },
      style: {
        borderTopColor: vars.colors.root.neutral[600],
      },
    },
    {
      variants: {
        theme: "light",
        variant: "primary",
      },
      style: {
        borderTopColor: vars.colors.root.neutral[600],
      },
    },
    {
      variants: {
        theme: "light",
        variant: "secondary",
      },
      style: {
        borderTopColor: vars.colors.root.neutral[0],
      },
    },
  ],
});
