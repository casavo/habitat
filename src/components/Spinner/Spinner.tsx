import { Circle, Container, Spin } from "./Spinner.css";

type Props = {
  size: "small" | "medium" | "large";
  theme: "light" | "dark";
  variant?: "primary" | "secondary" | "tertiary";
};

export const Spinner = ({ size, theme, variant }: Props) => {
  return (
    <div className={Container({ size })}>
      <div className={Circle({ size, theme, variant })}></div>
      <div className={Spin({ size, theme, variant })}></div>
    </div>
  );
};
