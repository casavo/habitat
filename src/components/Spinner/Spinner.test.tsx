import { render } from "@testing-library/react";
import { Spinner } from "./Spinner.tsx";
import { expect, describe } from "vitest";

describe("Spinner component", () => {
  it("renders", () => {
    const { container } = render(<Spinner size="medium" theme="dark" />);
    expect(container.firstChild).toMatchSnapshot();
  });
});
