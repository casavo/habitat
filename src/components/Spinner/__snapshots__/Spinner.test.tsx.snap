// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`Spinner component > renders 1`] = `
<div
  class="Spinner__gbijxg1 Spinner_size_medium__gbijxg3"
>
  <div
    class="Spinner_circle__gbijxg5 Spinner_circle_size_medium__gbijxgc Spinner_circle_theme_dark__gbijxg6"
  />
  <div
    class="Spinner_spin__gbijxgi Spinner_spin_size_medium__gbijxgp Spinner_spin_theme_dark__gbijxgm"
  />
</div>
`;
